ui/widgets/code_editor.go

```diff
@@ -175,28 +154,108 @@ func (c *CodeEditor) Layout(gtx layout.Context, theme *chapartheme.Theme, hint s
+							editorConf := &giovieweditor.EditorConf{
+								Shaper:          theme.Shaper,
+								TextColor:       theme.Fg,
+								Bg:              theme.Bg,
+								SelectionColor:  theme.TextSelectionColor,
+								TypeFace:        c.font.Font.Typeface,
+								TextSize:        unit.Sp(14),
+								LineHeightScale: 1.2,
+								ColorScheme:     "default",
+							}
+
+							c.editor.UpdateTextStyles(c.stylingText(c.editor.Text()))
+							return giovieweditor.NewEditor(c.editor, editorConf, hint).Layout(gtx)
 						})
 					}),
 				)
 			})
 		}),
 	)
 }
+
+var (
+	// List of Python keywords
+	keywords = []string{
+		"False", "await", "else", "import", "pass",
+		"None", "break", "except", "in", "raise",
+		"True", "class", "finally", "is", "return",
+		"and", "continue", "for", "lambda", "try",
+		"as", "def", "from", "nonlocal", "while",
+		"assert", "del", "global", "not", "with",
+		"async", "elif", "if", "or", "yield",
+	}
+	keyColor     = color.NRGBA{R: 255, G: 165, B: 0, A: 255}   // Orange
+	stringColor  = color.NRGBA{R: 42, G: 161, B: 152, A: 255}  // #2aa198
+	numberColor  = color.NRGBA{R: 42, G: 161, B: 152, A: 255}  // #2aa198
+	booleanColor = color.NRGBA{R: 128, G: 0, B: 128, A: 255}   // Purple
+	nullColor    = color.NRGBA{R: 128, G: 128, B: 128, A: 255} // Gray
+	commentColor = color.NRGBA{R: 88, G: 110, B: 117, A: 255}  // #586e75
+	// Define colors for different JSON elements
+	pythonKeywordsColor = color.NRGBA{R: 255, G: 165, B: 0, A: 255} // Orange
+
+	pythonPattern = regexp.MustCompile(`\b(` + strings.Join(keywords, "|") + `)\b`)
+
+	// Define regex patterns for different JSON elements
+	keyPattern                   = regexp.MustCompile(`"(\\\"|[^"])*"\s*:`)
+	stringPattern                = regexp.MustCompile(`"(\\\"|[^"])*"`)
+	numberPattern                = regexp.MustCompile(`\b\d+(\.\d+)?([eE][+-]?\d+)?\b`)
+	booleanPattern               = regexp.MustCompile(`\btrue\b|\bfalse\b`)
+	nullPattern                  = regexp.MustCompile(`\bnull\b`)
+	jsonSingleLineCommentPattern = regexp.MustCompile(`//.*`)
+	pythonCommentPattern         = regexp.MustCompile(`#.*`)
+)
+
+func (c *CodeEditor) stylingText(text string) []*giovieweditor.TextStyle {
+	if c.styledCode == text {
+		return c.styles
+	}
+
+	var styles []*giovieweditor.TextStyle
+
+	// Apply styles based on matches
+	applyStyles := func(re *regexp.Regexp, col color.NRGBA) {
+		matches := re.FindAllStringIndex(text, -1)
+		for _, match := range matches {
+			styles = append(styles, &giovieweditor.TextStyle{
+				Start: match[0],
+				End:   match[1],
+				Color: colorToOp(col),
+			})
+		}
+	}
+
+	// Always apply the JSON patterns
+	applyStyles(keyPattern, keyColor)
+	applyStyles(stringPattern, stringColor)
+	applyStyles(numberPattern, numberColor)
+	applyStyles(booleanPattern, booleanColor)
+	applyStyles(nullPattern, nullColor)
+	applyStyles(jsonSingleLineCommentPattern, commentColor)
+
+	if c.lang == CodeLanguagePython {
```

Comment:
Does code text mix JSON string and Python code? If not I would suggest using [Chroma](https://github.com/alecthomas/chroma ) for more comprehensive syntax support.

Reply:
editor will have one of kind each time. I tried to do it with Chroma but it does not return the `start` and `end` positions as it's needed in constructing the styles.

Reply:
I used chroma to do the same thing in another project. It may be helpful for you to have a look at this code snippet:
```go
var textStyles []*ed.TextStyle

	lexer := lexers.Get(lexerName)
	if lexer == nil {
		lexer = lexers.Fallback
	}

	lexer = chroma.Coalesce(lexer)

	// set style
	style := styles.Get(colorScheme)
	if style == nil {
		style = styles.Fallback
	}

	iterator, err := lexer.Tokenise(nil, string(source))
	if err != nil {
		return textStyles
	}

	offset := 0

	for _, token := range iterator.Tokens() {
		entry := style.Get(token.Type)

		textStyle := &ed.TextStyle{
			Start: offset,
			End:   offset + len([]rune(token.Value)),
		}

		if entry.Colour.IsSet() {
			textStyle.Color = colorToOp(uint32(entry.Colour))
		}

		// if entry.Background.IsSet() {
		// 	textStyle.Background = colorToOp(uint32(entry.Background))
		// }

		textStyles = append(textStyles, textStyle)
		offset = textStyle.End
		//log.Printf("color: %v, isset: %v, token: %s", entry.Colour, token.Type, token)

	}

	return textStyles
```

Reply:
Aha right, I see. 👍